pandas/tests/indexing/test_loc.py

```diff
@@ -690,6 +690,32 @@ def test_loc_setitem_consistency_slice_column_len(self):
         )
         tm.assert_series_equal(df[("Respondent", "Duration")], expected)
 
+    @pytest.mark.parametrize(
+        "dtype",
+        [
+            "datetime64",
+            "datetime64[ns]",
+            "datetime64[Y]",
+            "datetime64[M]",
+            "datetime64[D]",
+            "datetime64[h]",
+            "datetime64[m]",
+            "datetime64[s]",
+            "datetime64[ms]",
+            "datetime64[ns]",
+        ],
+    )
+    def test_loc_assign_non_ns_datetime(self, dtype):
+        df = pd.DataFrame(
+            {
+                "timestamp": [
+                    np.datetime64("2017-01-01 01:11:20"),
+                    np.datetime64("2017-01-01 02:01:30"),
+                ]
+            }
+        )
+        df.loc[:, "day"] = df.loc[:, "timestamp"].values.astype(dtype)
+
```

Comment:
Can you add an assert about the result here? `tm.assert_series_equal(df['day'], expected)`

Reply:
While I update the tests, I found a strange result. (Not sure this is intended.)

```python
df = pd.DataFrame({'timestamp': [np.datetime64('2017-01-01 01:02:23'), np.datetime64('2018-11-12 12:34:12')]})
df['year'] = df.loc[:, 'timestamp'].values.astype('datetime64[Y]')
df['month'] = df.loc[:, 'timestamp'].values.astype('datetime64[M]')
df.loc[:, 'day'] = df.loc[:, 'timestamp'].values.astype('datetime64[D]')
```
result:
```
            timestamp        day      month       year
0 2017-01-01 01:02:23 2017-01-01 2017-01-01 2017-01-01
1 2018-11-12 12:34:12 2018-11-12 2018-11-01 2018-01-01
```
`astype` result is okay.

```python
df.loc[:, 'timestamp'].values.astype('datetime64[Y]')
```
result:
```
array(['2017', '2018'], dtype='datetime64[Y]')
```

Reply:
Without running the test in master, it isn't obvious to the reader what is being tested.  I suggest something like:

```
@pytest.mark.parametrize("unit", ...)
def test_loc_setitem_datetime64_non_ns_converts(self, unit):
     # GH#27928 a few words about how this used to break
    df = DataFrame(
            {
                "timestamp": [
                    np.datetime64("2017-02-11 12:41:29"),
                    np.datetime64("1991-11-07 04:22:37"),
                ]
            }
        )
        df[unit] = df["timestamp"].values.astype("datetime64[{unit}]".format(unit=unit))
        df.loc[:, unit] = df["timestamp"].values.astype("datetime64[{unit}]".format(unit=unit))
```